lms/djangoapps/courseware/tests/test_views.py

```diff
@@ -307,6 +307,89 @@ def test_index_no_visible_section_in_chapter(self):
+        """
+        User will have staff access and creates the course
+        """
+        self.user = GlobalStaffFactory.create()
+        self.client.login(username=self.user.username, password='test')
+
+        # create the course
+        course = CourseFactory.create()
+        return course
+
+    def _create_url_for_enroll_staff(self, course):
+        """
+        creates the courseware url and enroll staff url
+        """
+        # create the _next parameter
+        courseware_url = reverse('courseware', kwargs={
+            'course_id': unicode(course.id)}) + '?activate_block_id=test_block_id'
+
+        # create the url for enroll_staff view
+        enroll_staff_url = "{enroll_staff_url}?next={courseware_url}".format(
+            enroll_staff_url=reverse('enroll_staff', kwargs={'course_id': unicode(course.id)}),
+            courseware_url=courseware_url
+        )
+        return enroll_staff_url
+
+    def test_redirection_unenrolled_staff(self):
+        """
+        Test that staff is not redirected to the 'about' page when visiting an unregistered course
+        """
+        course = self._create_staff_user_and_course()
+        enroll_staff_url = self._create_url_for_enroll_staff(course)
+
+        # Check the GET method
+        response = self.client.get(enroll_staff_url)
+        self.assertEqual(response.status_code, 200)
+        response_content = response.content
+        self.assertIn('Enroll', response_content)
+        self.assertIn("dont_enroll", response_content)
+
+    @ddt.data(
+        ({'enroll': "Enroll"}, 'courses/{}/courseware?activate_block_id=test_block_id'),
+        ({'dont_enroll': "Don't enroll"}, '/courses/{}/about'))
+    @ddt.unpack
+    def test_enroll_staff_redirection(self, data, expected_url_template):
+        """
+        Verify unenrolled staff is redirected to correct url.
+        """
+        course = self._create_staff_user_and_course()
+        enroll_staff_url = self._create_url_for_enroll_staff(course)
+        response = self.client.post(enroll_staff_url, data=data)
+
+        # Here we check the status code 302 because of the redirect
+        self.assertEqual(response.status_code, 302)
+        self.assertRedirects(response, expected_url_template.format(unicode(course.id)))
+
+    def test_enroll_staff_redirection_invalid_data(self):
+        """
+        If try to post with invalid data pattern then 404 will return
+        """
+        course = self._create_staff_user_and_course()
+        enroll_staff_url = self._create_url_for_enroll_staff(course)
+        response = self.client.post(enroll_staff_url, data={'test': "test"})
+        self.assertEqual(response.status_code, 404)
+
+    def test_courseware_redirection(self):
+        """
+        Test that staff is redirected to the 'enroll_staff' page when visiting  Courseware url
+        """
+        course = self._create_staff_user_and_course()
+        courseware_url = reverse('courseware', kwargs={
```

Comment:
I would do the following. Change `_create_url_for_enroll_staff` to return both `courseware_url` and `enroll_url`. Then, this test can do something like:

```python
self._create_global_staff_user()
courseware_url, enroll_url = self._create_url_for_enroll_staff()
response = self.client.get(courseware_url)
self.assertRedirects(response, enroll_url)
response = self.post(enroll_url, data={"enroll", "Enroll"})
self.assertRedirects(response, courseware_url)
self.assertTrue(CourseEnrollment.is_enrolled(self.global_staff, self.course.id))
```
Note that I haven't actually tested this out, but the idea is to test the whole workflow